Compression/src/CompressDetStepMCs_module.cc

```diff
@@ -0,0 +1,638 @@
+  keep_size += _simParticlesToKeep[i_product_id].size();
+  compressSimParticleCollection(_newSimParticlesPID, _newSimParticleGetter, *oldSimParticles, simPartSelector, *_newSimParticles);
+
+  // Fill out the SimParticleRemapping
+  for (const auto& i_keptSimPart : _simParticlesToKeep[i_product_id]) {
+    cet::map_vector_key oldKey = cet::map_vector_key(i_keptSimPart.key());
+    _simPtrRemap[i_keptSimPart] = art::Ptr<mu2e::SimParticle>(_newSimParticlesPID, oldKey.asUint(), _newSimParticleGetter);
+    if (_conf.debugLevel()>0) {
+      std::cout << "Compressing SimParticle " << i_keptSimPart << " --> " << _simPtrRemap[i_keptSimPart] << std::endl;
+    }
+  }
+  if (keep_size != _newSimParticles->size()) {
+    throw cet::exception("CompressDetStepMCs") << "Number of SimParticles in output collection (" << _newSimParticles->size() << ") does not match the number of SimParticles we wanted to keep (" << keep_size << ")" << std::endl;
+  }
+  if (_simParticleCompressionLevel == mu2e::CompressionLevel::kNoCompression) {
+    if (_newSimParticles->size() != oldSimParticles->size()) {
+      throw cet::exception("CompressDetStepMCs") << "Number of SimParticles in output collection (" << _newSimParticles->size() << ") does not match the number of SimParticles in the input collection (" << oldSimParticles->size() << ") even though no compression has been requested (simParticleCompressionLevel = \"" << _simParticleCompressionLevel.name() << "\")" << std::endl;
+    }
+  }
+
+  // If we asked for the genealogy to be compressed, we will now end up with some missing links
+  // add them back as truncated SimParticles
+  if (_keepNGenerations >= 0) {
+    // Go through the particles we are keeping and see if any parents are not there
+    for (const auto& i_keptSimPart : _simParticlesToKeep[i_product_id]) {
+
+      bool addedTruncated = false; // want to keep track of this because we don't want to add a truncated particle for every particle in the genealogy
+      int n_truncated_added = 0;
+      cet::map_vector_key truncKey;
+      art::Ptr<mu2e::SimParticle> truncSimPtr;
+      SimParticle* truncated = 0;
+
+      art::Ptr<mu2e::SimParticle> i_childPtr = i_keptSimPart;//art::Ptr<mu2e::SimParticle>(_newSimParticlesPID, i_newSimParticle.first.asUint(), _newSimParticleGetter);
+      art::Ptr<mu2e::SimParticle> i_parentPtr = i_childPtr->parent();
+      while (i_parentPtr) {
+        if (_simPtrRemap.find(i_parentPtr) == _simPtrRemap.end()) { // the parent is not in the output collection
+          if (_conf.debugLevel()>0) {
+            std::cout << "SimParticle " << i_parentPtr << " is not in output collection because it has been compressed away by genealogy compression" << std::endl;
+          }
+
+          // If we haven't added a truncated particle in this genealogy
+          if (!addedTruncated) {
+            // Add a truncated particle
+            truncated = new SimParticle();
```

Comment:
It looks like a single truncated particle is shared by all truncated genealogies.  It also appears it has no content.  Did you consider creating a unique truncated particle for each genealogy, including some content (like the ultimate parent, or the ultimate gen particle)?

Reply:
There should be a truncated particle for each genealogy branch. The only information I've set in the truncated particle is the parent ptr. I thought about adding a unique process code so that users can identify these truncated particle. I will bring this up to discuss in my talk. Also, the ultimate parent is always kept so that information should be safe.